p2p/transport/webrtc/datachannel.go

```diff
@@ -0,0 +1,282 @@
+package libp2pwebrtc
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"os"
+
+	// "io"
+	"net"
+
+	"sync"
+	"time"
+
+	"sync/atomic"
+
+	"github.com/libp2p/go-libp2p/core/network"
+	"github.com/pion/webrtc/v3"
+
+	pb "github.com/libp2p/go-libp2p/p2p/transport/webrtc/pb"
+)
+
+var _ network.MuxedStream = &dataChannel{}
+
+const defaultReadBufferLen = 4096
+
+// Package pion detached data channel into a net.Conn
+// and then a network.MuxedStream
+type dataChannel struct {
+	// TODO: Are these circular references okay?
+	channel       *webrtc.DataChannel
+	laddr         net.Addr
+	raddr         net.Addr
+	readDeadline  *deadline
+	writeDeadline *deadline
+
+	closeWriteOnce sync.Once
+	closeReadOnce  sync.Once
+	resetOnce      sync.Once
+
+	remoteWriteClosed uint32
+	localWriteClosed  uint32
+
+	remoteReadClosed uint32
+	localReadClosed  uint32
+
+	ctx        context.Context
+	cancel     context.CancelFunc
+	m          sync.Mutex
+	readBuf    bytes.Buffer
+	readSignal chan struct{}
+}
+
+func newDataChannel(
+	channel *webrtc.DataChannel,
+	pc *webrtc.PeerConnection,
+	laddr, raddr net.Addr) *dataChannel {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	result := &dataChannel{
+		channel:       channel,
+		laddr:         laddr,
+		raddr:         raddr,
+		readDeadline:  newDeadline(),
+		writeDeadline: newDeadline(),
+		ctx:           ctx,
+		cancel:        cancel,
+		readSignal:    make(chan struct{}),
+	}
+	result.readBuf.Grow(defaultReadBufferLen)
+
+	channel.OnMessage(result.handleMessage)
+
+	return result
+}
+
+func (d *dataChannel) handleMessage(msg webrtc.DataChannelMessage) {
+	if msg.IsString {
+		log.Warnf("received string message")
+		return
+	}
+
+	var pbmsg pb.Message
+	if err := pbmsg.Unmarshal(msg.Data); err != nil {
+		log.Warnf("could not unmarshal protobuf message")
+		return
+	}
+
+	if !d.isRemoteWriteClosed() && !d.isLocalReadClosed() {
+		d.m.Lock()
+		d.readBuf.Write(pbmsg.GetMessage())
+		// n, err := d.readBuf.Write(pbmsg.GetMessage())
+		// log.Warnf("wrote %d bytes to buffer, msg size: %d: %v", n, len(pbmsg.GetMessage()), err)
+		d.m.Unlock()
+		select {
+		case d.readSignal <- struct{}{}:
+		default:
+		}
+	}
+
+	if pbmsg.Flag != nil {
+		switch pbmsg.GetFlag() {
+		case pb.Message_FIN:
+			atomic.StoreUint32(&d.remoteWriteClosed, 1)
+			select {
+			case <-d.readSignal:
+			default:
+				close(d.readSignal)
+			}
+
+		case pb.Message_STOP_SENDING:
+			atomic.StoreUint32(&d.remoteReadClosed, 1)
+		case pb.Message_RESET:
+			log.Errorf("remote reset")
+			d.Close()
```

Comment:
@ckousik what is to be done about this?

Reply:
Code has completely changed since, so a new review in future will need to indicate if this is fixed / fine or not. For now can be ignored I am afraid.

Reply:
This comment can be resolved as such